ui/models.py

```diff
@@ -66,24 +69,85 @@ def s3_key(self):
+        """
+        Check the encode job status & if not complete, update the status via a query to AWS.
+
+        Returns:
+            str: The video's latest encode job status
+
+        """
+        if self.status in ['Complete', 'Error']:
+            return self.status
+        else:
+            job = get_et_job(self.encode_jobs.latest("created_at").id)
+            if job['Status'] == 'Complete':
+                self.process_transcode_results(job)
+            self.status = job['Status']
+            self.save()
+            return self.status
+
+    def transcode_key(self, preset=None):
+        """
+        Get the S3 key for a video playlist file
+
+        Args:
+            encoding(str): The encoding preset to use (should be one included in settings.ET_PRESET_IDS)
+
+        Returns:
+            str: The S3 key to used for the encoded file.
+        """
+        original_s3_key = self.videofile_set.get(encoding='original').s3_object_key
+        if not preset:
+            return original_s3_key
+        output_template = 'transcoded/{}_{}'
+        basename, _ = os.path.splitext(original_s3_key)
+        return output_template.format(basename, preset)
+
+    def process_transcode_results(self, job):
+        """
+        Create VideoFile and VideoThumbnail objects for a Video based on AWS ET job output
+
+        Args:
+            job(JSON): JSON representation of AWS ET job output
+        """
+
+        for playlist in job['Playlists']:
+            VideoFile.objects.get_or_create(
+                bucket_name=settings.VIDEO_S3_TRANSCODE_BUCKET,
+                s3_object_key='{}.m3u8'.format(playlist['Name']),  # This assumes HLS encoding
+                encoding=EncodingNames.HLS,
+                preset_id=','.join([output['PresetId'] for output in job['Outputs']]),
+                video=self
+            )
+        for output in job['Outputs']:
+            if 'ThumbnailPattern' in output:
+                thumbnail_pattern = output['ThumbnailPattern'].replace("{count}", "")
+                preset = get_et_preset(output['PresetId'])
+                s3 = boto3.resource("s3")
+                bucket = s3.Bucket(settings.VIDEO_S3_THUMBNAIL_BUCKET)
+                for thumb in bucket.objects.filter(Prefix=thumbnail_pattern):
+                    VideoThumbnail.objects.get_or_create(
+                        video=self,
+                        bucket_name=settings.VIDEO_S3_THUMBNAIL_BUCKET,
+                        s3_object_key=thumb.key,
+                        preset_id=output['PresetId'],
+                        max_height=int(preset['Thumbnails']['MaxHeight']),
+                        max_width=int(preset['Thumbnails']['MaxWidth'])
+                    )
+
     def __str__(self):
         return self.title or "<untitled video>"
 
     def __repr__(self):
         return '<Video: {self.title!r} {self.s3_subkey!r}>'.format(self=self)
 
 
-class VideoFile(models.Model):
+class VideoS3Mixin(object):
```

Comment:
How about keeping this a subclass of `models.model`, marking it an abstract model, and retaining these fields: `created_at`, `s3_object_key`, `s3_object_key` and the method `delete_s3_objects` as implemented in `VideoThumbnail`.